xcvm/cosmwasm/contracts/interpreter/src/contract.rs

```diff
@@ -0,0 +1,274 @@
+use cosmwasm_std::{
+	entry_point, to_binary, CosmosMsg, DepsMut, Env, MessageInfo, QueryRequest, Response, StdError,
+	WasmQuery,
+};
+use serde::Serialize;
+
+use crate::{
+	error::ContractError,
+	msg::{ExecuteMsg, InstantiateMsg, XCVMProgram},
+	state::{Config, CONFIG},
+};
+use cw20::{BalanceResponse, Cw20Contract, Cw20ExecuteMsg, Cw20QueryMsg};
+use xcvm_asset_registry::msg::{GetAssetContractResponse, QueryMsg as AssetRegistryQueryMsg};
+use xcvm_core::{Amount, Funds, Instruction, NetworkID};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+	deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	msg: InstantiateMsg,
+) -> Result<Response, StdError> {
+	let registry_address = deps.api.addr_validate(&msg.registry_address)?;
+
+	let config = Config { registry_address };
+
+	CONFIG.save(deps.storage, &config)?;
+
+	Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+	deps: DepsMut,
+	env: Env,
+	info: MessageInfo,
+	msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+	match msg {
+		ExecuteMsg::Execute { program } => interpret_program(deps, env, info, program),
+	}
+}
+
+pub fn interpret_program(
+	mut deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	program: XCVMProgram,
+) -> Result<Response, ContractError> {
+	let mut response = Response::new();
+
+	for instruction in program.instructions {
+		response = match instruction {
+			Instruction::Call { encoded } => interpret_call(encoded, response),
+			Instruction::Spawn { network, salt, assets, program } =>
+				interpret_spawn(network, salt, assets, program, response),
+			Instruction::Transfer { to, assets } =>
+				interpret_transfer(&mut deps, to, assets, response),
+		}?;
+	}
+	Ok(response)
+}
+
+pub fn interpret_call(encoded: Vec<u8>, response: Response) -> Result<Response, ContractError> {
+	let cosmos_msg: CosmosMsg =
```

Comment:
This is actually pretty cool. We used to expect payload for `WasmMsg::Execute`. And we were using a custom payload format such as:
```rust
struct Payload {
    address: String,
    msg: String
}
```
which is not very nice. Now we expect a `CosmosMsg`. This means that users can do any kind of stuff that cosmwasm environment support such as transfer native tokens, instantiate token, custom messages, etc.